contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,151 @@
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND mat_$current_url = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+Looking at `system.query_log`, the new query:
+
+- Took 980ms (**71%/3.4x improvement**)
+- Read 14.36 GiB from disk (**81%/5x improvement improvement**)
+
+The wins are even more magnified if more than one property filter is used at a time.
+
+## Backfilling efficiently
+
+Using `OPTIMIZE TABLE` after adding columns is often not a good idea, since it will involve a lot of I/O as the whole table gets rewritten.
+
+As of writing, there's a feature request on [Github](https://github.com/ClickHouse/ClickHouse/issues/27730) for adding specific commands for materializing specific columns on ClickHouse data parts.
+
+Alternatively, `DEFAULT` type columns can be used for backfilling data.
+
+```sql
+ALTER TABLE events
+ALTER COLUMN mat_$current_url
+VARCHAR DEFAULT JSONExtractString(properties_json, '$current_url');
+
+ALTER TABLE events UPDATE mat_$current_url = mat_$current_url WHERE timestamp >= '2021-08-01';
+
+-- Wait for mutations to finish before running this
+ALTER TABLE events
+ALTER COLUMN mat_$current_url
+VARCHAR MATERIALIZED JSONExtractString(properties_json, '$current_url');
+```
+
+This will compute and store only the `mat_$current_url` in our time range, making it much cheaper.
+
+Be aware though that this will:
+1. Break your `INSERT` statements if you don't specify column names explicitly
+2. Alter the behavior of `SELECT *` queries
+
+## Usage at PostHog
+
+PostHog as an analytics tool allows users to slice and dice their data in many ways across huge time ranges and datasets. This also means that performance is key when investigating things - but also that we currently do nearly no preaggregation.
+
+Rather than materialize all columns, we built a solution that looks at recent slow queries using `system.query_log`, determines which properties need materializing from there, and backfills the data on a weekend. This works well because not every query needs optimizing and a relatively small subset of properties make up most of what’s being filtered on by our users.
+
+You can find the code for this [here](https://github.com/PostHog/posthog/blob/c23704b3909ae8ebb827e6a43453e32b3d3487bd/ee/clickhouse/materialized_columns/analyze.py#L42-L119) and [here](https://github.com/PostHog/posthog/blob/c23704b3909ae8ebb827e6a43453e32b3d3487bd/ee/clickhouse/materialized_columns/columns.py#L37-L130).
+
+After materializing our top 100 properties and updating our queries, we analyzed slow queries (>3 seconds long). **The average improvement in our query times was 55%, with 99th percentile improvement being 25x.**
+
+As a product, we're only scratching the surface of what ClickHouse can do to power product analytics. If you're interested in helping us with these kinds of problems, [we're hiring](https://posthog.com/careers)!
```

Comment:
I would add further reading links here to the ClickHouse docs and Altinity docs if you have any as references